src/Microsoft.AspNetCore.Mvc.Core/ControllerAttribute.cs

```diff
@@ -0,0 +1,17 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNetCore.Mvc
+{
+    /// <summary>
+    /// Indicates that the type and any derived types that this attribute is applied to
+    /// are considered a controller by the default controller discovery mechanism, unless
+    /// <see cref="NonControllerAttribute"/> is applied to any type in the hierarchy.
```

Comment:
@dougbu I've explictly been very simplistic about this behavior. NonController always wins over Controller. It is easier behavior to understand for our users. With other workitems that I'm readying for PR, adding or removing controllers to the list of discovered controller types will be really easy, that's why I chose to keep our convention simple and that would be the recommended way for our customers to tweak the results of discovery.